### PR TITLE
SparkR Support

### PR DESCRIPTION
**Pull request by ifilonenko**
## What changes were proposed in this pull request?

Initial Spark R support

## How was this patch tested?

- [x] Initial submission step 
- [x] Unit Tests
- [x] Docker files (tested)
- [x] Integration Tests

### CODE REVIEW

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigurationStepsOrchestrator.scala [hunk 1]
```diff
@@ -71,6 +76,7 @@ private[spark] class DriverConfigurationStepsOrchestrator(
         .map(_.split(","))
         .getOrElse(Array.empty[String]) ++
         additionalMainAppPythonFile.toSeq ++
+        additionalMainAppRFile.toSeq ++
```
Comment (ifilonenko) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigurationStepsOrchestrator.scala:80]: Important here that, similar to Python Primary Resource, that the R File is distributed via --files.

File: resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/driver-r/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+# If this docker file is being used in the context of building your images from a Spark distribution, the docker build
+# command should be invoked from the top level directory of the Spark distribution. E.g.:
+# docker build -t spark-driver-f:latest -f dockerfiles/driver-f/Dockerfile .
+
+ADD examples /opt/spark/examples
+ADD R /opt/spark/R
+
+RUN apk add --no-cache R && \
+    rm -r /root/.cache
```
Comment (ifilonenko) [inline resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/driver-r/Dockerfile]: Open here for any recommendations?

**Comment by ifilonenko** [comment]
Upon merging, this PR closes #506

**Comment by ifilonenko** [comment]
To run integration tests in a proper R environment, it is required that R_HOME is defined in the testing environment which means everyone would need to install R. Is that something that would be an issue or something that can be assumed if someone is building out a full dev environment for Spark?  @foxish @erikerlandson

**Comment by foxish** [comment]
Hmm, interesting. The submitting node has an R dependency? or the driver?

**Comment by ifilonenko** [comment]
The R dependency is because there is a need to mimic the make-distribution environment in `target/docker/R` so that when we run `ADD R opt/spark/R` it is already packaged in the Docker environment; this is similar to how we setup PySpark.

**Comment by ifilonenko** [comment]
@ssuchter @varunkatta Integration test will pass after R is installed and R_HOME is defined in the jenkins environment. 

PR is otherwise ready for review

**Comment by ifilonenko** [comment]
rerun integration tests please

**Comment by ifilonenko** [comment]
rerun integration tests please

**Comment by ifilonenko** [comment]
rerun integration tests please

**Comment by ifilonenko** [comment]
PR is ready for review @foxish @erikerlandson

**Comment by liyinan926** [comment]
@ifilonenko you also need to update `sbin/build-push-docker-images.sh` to add the new images.

**Comment by ifilonenko** [comment]
rerun unit tests please

**Comment by ifilonenko** [comment]
Ready for merging to branch-2.2 unless any other concerns @erikerlandson @foxish @liyinan926

**Comment by liyinan926** [comment]
LGTM. Thanks for the work!

**Comment by ifilonenko** [comment]
rerun unit tests please

**Review by felixcheung — APPROVED** [review]

**Comment by ifilonenko** [comment]
Unless there are any further comments, I think this is ready to merge

**Comment by ifilonenko** [comment]
All set to merge? @foxish @erikerlandson @liyinan926

**Comment by liyinan926** [comment]
I think it's all good.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala [hunk 3]
```diff
@@ -101,6 +101,32 @@ private[spark] class KubernetesSuite extends SparkFunSuite with BeforeAndAfter {
     runPySparkPiAndVerifyCompletion(PYSPARK_PI_CONTAINER_LOCAL_FILE_LOCATION, Seq.empty[String])
   }
 
+  test("Run SparkR Job on file locally") {
+    assume(testBackend.name == MINIKUBE_TEST_BACKEND)
+
+    launchStagingServer(SSLOptions(), None)
```
Comment (mccheah) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:107]: Don't think we need a staging server if the file is in the image.

Comment (mccheah) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:107]: Actually I just think the test is misnamed - looks like this test is shipping the file while the test below expects the file to be on the container.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/docker/SparkDockerImageBuilder.scala [hunk 4]
```diff
@@ -76,21 +78,34 @@ private[spark] class SparkDockerImageBuilder
...
-    builder.redirectErrorStream(true) // Ugly but needed for stdout and stderr to synchronize
-    val process = builder.start()
+    python_builder.directory(new File(DOCKER_BUILD_PATH.toFile, "python"))
+    python_builder.redirectErrorStream(true) // Ugly but needed for stdout and stderr to synchronize
+    val process = python_builder.start()
     new RedirectThread(process.getInputStream, System.out, "redirect output").start()
-    val exitCode = process.waitFor()
-    if (exitCode != 0) {
-      logInfo(s"exitCode: $exitCode")
+    val exitCodePython = process.waitFor()
```
Comment (mccheah) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/docker/SparkDockerImageBuilder.scala:87]: I've always been wondering - would it be possible to do this bootstrapping from Maven and not in the test code? Seems like this should be an environment step. Docker images should theoretically be built at the Maven step as well but we know that this is harder to do.

**Comment by felixcheung** [comment]
hi - what's left for this work? this https://github.com/apache-spark-on-k8s/spark/pull/507#discussion_r143595440?

**Comment by mccheah** [comment]
I left some comments. Also would like to see this tested in a production environment, but maybe we can just merge it and follow up as feedback comes in.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/docker/SparkDockerImageBuilder.scala (hunk 4 above)
Comment (ifilonenko) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/docker/SparkDockerImageBuilder.scala:87]: It technically is possible, but as we discussed for PySpark. I wasn't able to figure that out. Any recommendations would be helpful.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala (hunk 3 above)
Comment (ifilonenko) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:107]: `src/test/R/dataframe.R` is what we are pushing up to the RSS, which is why it is being launched. Hmm, isn't that file locally stored? What would be the correct naming convention?

**Comment by ifilonenko** [comment]
@mccheah as per your last comment. Is this okay to merge then?

**Comment by mccheah** [comment]
Can merge when CI passes - I just updated the branch.

**Comment by ifilonenko** [comment]
ready for merging: @foxish

**Comment by foxish** [comment]
Nvm. Please ignore last comment. That was in the merge commit.

**Pull request merged**